baseline/services.py

```diff
@@ -481,25 +483,40 @@ def predict(self, tokens, **kwargs):
             examples = unfeaturized_examples
 
         outcomes = self.model.predict(examples)
-        return self.format_output(outcomes, tokens_batch=tokens_batch, label_field=label_field)
+        return self.format_output(outcomes, tokens_batch=tokens_batch, label_field=label_field, vectorized_examples=examples)
 
-    def format_output(self, predicted, tokens_batch=None, label_field='label', **kwargs):
+    def format_output(self, predicted, tokens_batch=None, label_field='label', vectorized_examples=None, **kwargs):
+        """This code got very messy dealing with BPE/WP outputs."""
         assert tokens_batch is not None
+        assert vectorized_examples is not None
         outputs = []
+        # Pick a random non-lengths key from the vectorized input, if one input was BPE'd they all had to be to stay aligned
+        key = [k for k in vectorized_examples if not k.endswith("_lengths")][0]
+        # Get the subword start signal from the vectorizer's tokenizer. If there is none we return None
+        subword_start = getattr(getattr(self.vectorizers[key], "tokenizer", None), "subword_start", None)
+        # For each item in a batch
         for i, outcome in enumerate(predicted):
             output = []
-            j = 0
-            for token in tokens_batch[i]:
-                new_token = dict()
-                new_token.update(token)
-                label = outcome[j] if self.return_labels else self.label_vocab[outcome[j].item()]
-                while label == Offsets.VALUES[Offsets.PAD] and j < len(outcome) - 1:
-                    j += 1
-                    label = outcome[j] if self.return_labels else self.label_vocab[outcome[j].item()]
+            # Extract the vectorized example for this batch element and the key we are choosing
+            vectorized_example = vectorized_examples[key][i]
+            # Convert back into strings, these will now be broken into subwords
+            tokenized_text = [self.rev_vocab[key][t] for t in vectorized_example]
+            new_outcome = []
+            # Loop through all the tokens and collect the labels for the start of subwords
+            for j, t in enumerate(tokenized_text):
+                # If the subword start is None we are a normal vectorizer and want to collect everything
+                # If we start with the subword start we are part of a broken token and our label doesn't matter
+                # If we are the special tags from pre-training our labels don't matter
+                if subword_start is None or (not t.startswith(subword_start) and t not in ("[CLS]", "[SEP]", "[PAD]")):
```

Comment:
i think the SoC here is somewhat lacking.  wouldnt it be better to ask the vectorizer what its special tokens are?  RN, these are defined as string literals in the middle of the code and these types of items also happen in the vectorizers, why not place them all in the same area?

Reply:
BPE sub-words have @@ at the end: 'Ro@@ asted barr@@ am@@ un@@ di fish', 'Centr@@ ally managed over a cli@@ ent-@@ server architecture']

Reply:
possibly another property in the vectorizer could abstract (`is_leader_token`?)

Reply:
Yeah I just noticed that the bpe has a trailing signal instead of a starting one.

Having that be different makes things more complex.

What if the vectorizers had a method that took in a tokenized string (or they could rerun the tokenization but that would be wasteful) and output the indices that were valid tokens?

Reply:
yeah, that would work I guess